keps/sig-storage/3333-reconcile-default-storage-class/README.md

```diff
@@ -0,0 +1,666 @@
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+    - [ ] e2e Tests for all Beta API Operations (endpoints)
+    - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+    - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+    - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+We intend to change behaviour of default storage class assignment to be
+retroactive for existing *unbound* persistent volume claims without any storage
+class assigned. This changes the existing Kubernetes behaviour slightly, which
+is further described in sections below.
+
+A PVC with `pvc.spec.storageClassName=nil` will **always** get a default storage
+class, regardless if the storage class is created before or after the PVC is
+created.
+
+## Motivation
+
+When user needs to provision a storage they create a PVC to request a volume. 
+A control loop looks for any new PVCs and based on current state of the 
+cluster the volume will be provided using one of the following methods:
+
+* Static provisioning - PVC did not specify any storage class and there is
+  already an existing PV that can be bound to it. Alternatively users can set
+  `pvc.spec.storageClassName=""` to disable dynamic provisioning explicitly.
+* Dynamic provisioning - there is no existing PV that could be bound but PVC 
+  did specify a storage class or there is exactly one storage class in the 
+  cluster marked as default.
+
+Considering the "normal" operation described above there are additional
+cases that can be problematic:
+
+1. It’s hard to mark a different SC as the default one. Cluster admin can
+   choose between two bad solutions:
+
+    1. Cluster has two default SCs for a short time, i.e. admin marks the new
```

Comment:
We should change this semantic - while there are two, pick one randomly or heuristically.  In normal use it's the same thing as losing a race condition.

E.g., in real-time you have one user changing the default and another user creating a PVC at the same time.  Depending on network, CPU scheduling, etc (totally non-deterministic) the API server might see either of those operations first.  The PVC will either win the race and get the old default or it will lose the race and get the new default.

So whichever one you pick, as long as it is one of the two defaults, it is a correct choice!